renku/domain_model/template.py

```diff
@@ -586,3 +588,73 @@ def update(self, template: Template):
         self.metadata["__template_id__"] = template.id
         self.metadata["__automated_update__"] = template.allow_update
         self.immutable_files = template.immutable_files
+
+
+def find_renku_section(lines: List[str]) -> Tuple[int, int]:
+    """Return start and end line numbers of the Renku-specific section."""
+    start = end = -1
+    for index, line in enumerate(lines):
+        if line.startswith("#        Renku-specific section - DO NOT MODIFY        #"):
+            start = index
+        elif line.endswith("#              End Renku-specific section              #"):
+            end = index
+            break
+
+    return start, end
+
+
+def get_renku_section_from_dockerfile(content: str) -> str:
+    """Return the Renku-specific section of the Dockerfile or the whole Dockerfile if it doesn't exist."""
+    lines = [line.rstrip() for line in content.splitlines()]
+    start, end = find_renku_section(lines)
+
+    if 0 <= start < end:
+        lines = lines[start:end]
+        lines = [line for line in lines if line]  # NOTE: Remove empty lines
+        return "\n".join(lines)
+    else:
+        return content
+
+
+def calculate_dockerfile_checksum(
+    *, dockerfile: Optional[Path] = None, dockerfile_content: Optional[str] = None
+) -> str:
+    """Calculate checksum for the given file or content.
+
+    NOTE: We ignore empty lines and whitespace characters at the end of the lines when calculating Dockerfile checksum
+    if it has Renku-specific section markers.
+    """
+    if not dockerfile and not dockerfile_content:
+        raise errors.ParameterError("Either Dockerfile or its content must be passed")
+    elif dockerfile and dockerfile_content:
+        raise errors.ParameterError("Cannot pass both Dockerfile and its content")
+
+    content = dockerfile_content if dockerfile_content is not None else dockerfile.read_text()  # type: ignore
+    renku_section = get_renku_section_from_dockerfile(content)
+    return hash_string(renku_section)
+
+
+def update_dockerfile_content(source: Path, destination: Path) -> None:
+    """Update the Renku-specific section of the destination Dockerfile with the one from the source Dockerfile."""
+    source_lines = [line.rstrip() for line in source.read_text().splitlines()]
+    source_start, source_end = find_renku_section(source_lines)
+
+    destination_lines = [line.rstrip() for line in destination.read_text().splitlines()]
+    destination_start, destination_end = find_renku_section(destination_lines)
+
+    # NOTE: If source or destination Dockerfiles doesn't have Renku-specific section, we overwrite the whole file
```

Comment:
Is this potentially problematic from a user perspective? I.e. if I have an old project where I have done custom changes but I do not have the renku section then renku will tell me that I can upgrade automatically but then also wipe out my custom changes. But maybe I misunderstand.

Should we avoid updating altogether in this case?